src/main/navigation/navigation.c

```diff
@@ -2029,11 +2077,41 @@ static void updateDesiredRTHAltitude(void)
 {
     if (ARMING_FLAG(ARMED)) {
         if (!(navGetStateFlags(posControl.navState) & NAV_AUTO_RTH)) {
-            posControl.homeWaypointAbove.pos.z = RTHAltitude();
+            switch (navConfig()->general.flags.rth_alt_control_mode) {
+                case NAV_RTH_NO_ALT:
+                    posControl.rthState.rthInitialAltitude = posControl.actualState.abs.pos.z;
+                    posControl.rthState.rthFinalAltitude = posControl.actualState.abs.pos.z;
+                    break;
+
+                case NAV_RTH_EXTRA_ALT: // Maintain current altitude + predefined safety margin
+                    posControl.rthState.rthInitialAltitude = posControl.actualState.abs.pos.z + navConfig()->general.rth_altitude;
+                    posControl.rthState.rthFinalAltitude = posControl.actualState.abs.pos.z + navConfig()->general.rth_altitude;
+                    break;
+
+                case NAV_RTH_MAX_ALT:
```

Comment:
I think there is an issue with `updateDesiredRTHAltitude()` and the `NAV_RTH_MAX_ALT`, `NAV_RTH_AT_LEAST_ALT` and `NAV_RTH_AT_LEAST_ALT_LINEAR_DESCENT` altitude modes.

`updateDesiredRTHAltitude()` is called regularly (10Hz) by `updatePositionEstimator()` as soon as the position is valid and each time it is called and the aircraft is armed and not executing RTH `posControl.rthState.rthInitialAltitude` will be increased by `navConfig()->general.rth_altitude` in the case where `posControl.rthState.rthInitialAltitude + navConfig()->general.rth_altitude` is greater than `posControl.actualState.abs.pos.z` so as soon as the aircraft is lower than the set RTH altitude.

```
posControl.rthState.rthInitialAltitude = MAX(posControl.rthState.rthInitialAltitude + navConfig()->general.rth_altitude, posControl.actualState.abs.pos.z);
```

The fix is probably to replace `posControl.rthState.rthInitialAltitude` in the rvalues by `posControl.rthState.homePosition.pos.z`

Reply:
Well spotted, thanks!